theano/tensor/elemwise.py

```diff
@@ -870,6 +870,13 @@ def perform(self, node, inputs, output_storage):
             elif (not isinstance(variable, numpy.ndarray) or
                   variable.dtype != nout.dtype):
                 variable = numpy.asarray(variable, nout.dtype)
+                # The next line is needed for numpy 1.9. Otherwise
+                # there are tests that fail in DebugMode.
+                # Normally we would call theano.misc._asarray, but it
+                # is faster to inline the code. We know that the dtype
+                # are the same string, just different typenum.
+                if numpy.dtype(nout.dtype).num != variable.dtype.num:
+                    variable = variable.view(dtype=nout.dtype)
```

Comment:
This happen in a test in DebugMode with numpy 1.9.

This line show it ( I forgot which one in the batch caused this error).

THEANO_FLAGS=optimizer_excluding=local_useless_crossentropy_softmax_1hot_with_bias_dx_alloc:local_useless_inc_subtensor_alloc,on_shape_error=raise,on_opt_error=raise,,warn.argmax_pushdown_bug=False,warn.gpusum_01_011_0111_bug=False,warn.sum_sum_bug=False,warn.sum_div_dimshuffle_bug=False,warn.subtensor_merge_bug=False,,device=cpu,floatX=float64,unittests.rseed=11212,mode=DEBUG_MODE,DebugMode.check_strides=0,DebugMode.patience=3,DebugMode.check_preallocated_output=
theano/../bin/theano-nose
theano/tensor/tests/test_basic.py:{ErfcinvTester,ErfinvTester,TestInferShape,TestTensorInstanceMethods}
theano/tensor/tests/test_extra_ops.py:TestBinCountOp
theano/tensor/tests/test_nlinalg.py:test_qr_modes --pdb --pdb-failure -s

For elemwise of scalar, we receive an numpy scalar, so we call
numpy.asarray() on it with dtype 'int64' but it return a numtype of 9
(longlong) instead of 7 (int64). This show the problem. You can use my
software stack v4 in the lab.

>>> numpy.asarray([], dtype='longlong').dtype == 'int64'
True
>>> numpy.asarray([], dtype='longlong').dtype.num
9
>>> numpy.asarray([], dtype='int64').dtype.num
7




On Fri, Apr 24, 2015 at 1:47 PM, abergeron <notifications@github.com> wrote:

> In theano/tensor/elemwise.py
> <https://github.com/Theano/Theano/pull/2824#discussion_r29069579>:
>
> > @@ -870,6 +870,13 @@ def perform(self, node, inputs, output_storage):
> >              elif (not isinstance(variable, numpy.ndarray) or
> >                    variable.dtype != nout.dtype):
> >                  variable = numpy.asarray(variable, nout.dtype)
> > +                # The next line is needed for numpy 1.9. Otherwise
> > +                # there are tests that fail in DebugMode.
> > +                # Normally we would call theano.misc._asarray, but it
> > +                # is faster to inline the code. We know that the dtype
> > +                # are the same string, just different typenum.
> > +                if numpy.dtype(nout.dtype).num != variable.dtype.num:
> > +                    variable = variable.view(dtype=nout.dtype)
>
> I find the justification shaky.
>
> Do you mean that you have something like nout.dtype == 'float32' and
> variable.dtype.num == NPY_FLOAT32 and they are somehow not equal?
>
> If I misunderstood can you give me an example of a problem that fixes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/2824/files#r29069579>.
>